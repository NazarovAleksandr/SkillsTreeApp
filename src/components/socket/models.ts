import { IRune } from "../runes/models";

export interface SocketProps {
    rune: IRune;
    children: React.ReactNode;
    isRuneOver?: boolean;
    nodeId: string;
}