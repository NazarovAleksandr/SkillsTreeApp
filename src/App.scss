@import './variables.scss';

#root, body, html {
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    background-color: black;
}

.App {
    font-family: Arial, Helvetica, sans-serif;
    height: 100%;
    display: flex;
    margin: auto;
    max-width: 1650px;

    .layout {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        margin: 3%;

        .pages {
            flex-grow: 1;
            display: flex;
            overflow: auto;
        }
    }

    .menu {
        border-bottom: 2px solid $mainBorderColor;
        margin-bottom: 10px;
        text-align: center;

        > a {
            display: inline-block;
            margin-right: 20px;
            padding: 5px;
            color: #faa318;
            text-decoration: none;
            width: 100px;

            &:last-child {
                margin-right: 0;
            }

            &.active, &:hover {
                box-shadow: 0px 2px #faa318;
            }
        }
    }
}