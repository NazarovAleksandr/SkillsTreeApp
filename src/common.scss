@import './variables.scss';

.rune-socket {
    width: 64px;
    height: 64px;
    display: inline-block;
    border: 2px solid;
    border-radius: 40px;
    background: #171717;
    vertical-align: middle;
    box-shadow: 0 2px #000;
    margin: 8px;

    &.unique {
        color: $uniqueRarity;
    }

    &.common {
        color: $commonRarity;
    }

    &.rare {
        color: $rareRarity;
    }
}

.rune {
    &-image {
        width: 64px;
        height: 64px;
    }

    &.used {
        background-color: yellow;
    }
}

.block-title {
    border-bottom: 2px solid $mainBorderColor;
    margin-bottom: 10px;
    padding: 10px;
    color: #faa318;

    .title {
        padding: 6px;
        display: inline-block;
        text-transform: uppercase;
    }
}

.action {
    color: rgb(250, 163, 24);
    border: 1px solid #545454;
    display: inline-block;
    padding: 5px;
    box-shadow: 2px 2px #000;
    text-align: center;
    cursor: default;
    position: relative;
    user-select: none;

    &.disabled {
        color: gray;
    }

    &:hover:not(.disabled) {
        border-color: rgb(250, 163, 24);
        cursor: pointer;
    }
}

@keyframes drop-splash {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	100% {
		opacity: 0;
        transform: scale3d(1.1, 1.1, 1);
        width: 70px;
        height: 70px;
	}
}

@keyframes drop-splash-2 {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	50%, 100% {
        width: 70px;
        height: 70px;
		opacity: 0;
		transform: scale3d(1.2, 1.2, 1);
	}
}

@keyframes rotation {
    100% {
        transform: rotate(360deg);
    }
}