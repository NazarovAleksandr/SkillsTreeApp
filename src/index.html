<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8"/>
    <meta name="viewport" content="initial-scale=0.6, shrink-to-fit=yes">
    <title>Game proof-of-concept</title>
  </head>
  <body>
    <div id="root"></div>
</html>