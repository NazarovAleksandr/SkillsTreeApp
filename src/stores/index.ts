import * as magicSchools from './magicSchool';
import * as runes from './runes';
import * as skillTrees from './skilltrees';
import * as tooltipState from './tooltipState';

export {
    magicSchools,
    runes,
    skillTrees,
    tooltipState
}