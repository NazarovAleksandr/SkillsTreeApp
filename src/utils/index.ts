import * as randomizer from './randomizer';
import * as uids from './uids';
import * as constants from './constants';
import { images } from './images';

export {
    randomizer,
    uids,
    constants,
    images,
};
