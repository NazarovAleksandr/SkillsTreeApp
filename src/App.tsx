import * as React from 'react';
import './App.scss';
import './common.scss';
import { observer } from "mobx-react";
import AboutPage from './pages/about';
import UserSkillsPage from './pages/userSkills';
import { BrowserRouter as Router, Route, NavLink } from "react-router-dom";
import { TooltipContext } from './contexts/tooltip';
import * as stores from './stores';
import { Tooltip } from './components/tooltip/tooltip';

const tooltipStateStore = new stores.tooltipState.TooltipStateStore();
let basename = process.env.BASENAME || '/'

window.addEventListener('scroll', () => {
    tooltipStateStore.hide();
}, {
    capture: true,
    passive: true
});

const App = () => {
    return (
        <Router basename={basename}>
            <TooltipContext.Provider value={tooltipStateStore}>
                <div className="App">
                    <div className="layout">
                        <div className="menu">
                            <NavLink exact to="/" activeClassName='active'>About</NavLink>
                            <NavLink to="/skills" activeClassName='active'>Skills</NavLink>
                        </div>
                        <div className="pages">
                            <Route path="/skills" component={UserSkillsPage} />
                            <Route exact path="/" component={AboutPage} />
                        </div>
                    </div>
                </div>
            </TooltipContext.Provider>
            <Tooltip 
                position={tooltipStateStore.tooltipPosition}
                content={tooltipStateStore.tooltipContent}
                node={tooltipStateStore.hostNode}
                visible={tooltipStateStore.isVisible}
                mouseOutCallback={tooltipStateStore.mouseOutCallback}
            />
        </Router>
    );
}

export default observer(App);

/*

TODO

1) Рефакторинг

*/