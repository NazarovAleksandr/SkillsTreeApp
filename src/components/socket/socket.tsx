import * as React from 'react';
import { SocketProps } from './models';
import { constants } from '../../utils';

export class Socket extends React.PureComponent<SocketProps> {
    public socketRef = React.createRef<HTMLSpanElement>();

    public animationTimeout = 1000;

    public componentDidUpdate(prevProps: SocketProps) {
        const { rune, nodeId } = this.props;
        if (prevProps.rune !== rune && nodeId === prevProps.nodeId) {
            this.socketRef.current.classList.add(constants.splashAnimated);
            setTimeout(() => {
                this.socketRef.current.classList.remove(constants.splashAnimated);
            }, this.animationTimeout);
        }
    }

    public render() {
        const { children, rune, isRuneOver } = this.props;
        const socketClassName = `rune-socket ${isRuneOver ? 'drop-allowed' : ''} ${rune ? `${rune.rarity.toLowerCase()}-splash` : ''}`;
        return (
            <span className={socketClassName} ref={this.socketRef}>
                {children}
            </span>
        );
    }
}
