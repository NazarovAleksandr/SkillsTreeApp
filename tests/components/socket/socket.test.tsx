import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Socket } from '../../../src/components/socket/socket';
import {
    IRune, Rarity, RuneTypes, RunePropertyTypes,
} from '../../../src/components/runes/models';
import * as utils from '../../../src/utils';

describe('Socket', () => {
    test('Renders with child', () => {
        const rune: IRune = {
            id: '1',
            image: 'url',
            rarity: Rarity.common,
            type: RuneTypes.attack,
            properties: [
                { name: RunePropertyTypes.Evade, value: 10 },
            ],
        };

        const nodeId = '1';
        const component = shallow(<Socket rune={rune} nodeId={nodeId}>test</Socket>);

        expect(component).toMatchSnapshot('socket with child');
    });
    test('Renders with rune over', () => {
        const rune: IRune = {
            id: '1',
            image: 'url',
            rarity: Rarity.common,
            type: RuneTypes.attack,
            properties: [
                { name: RunePropertyTypes.Evade, value: 10 },
            ],
        };

        const nodeId = '1';
        const component = shallow(<Socket rune={rune} nodeId={nodeId} isRuneOver>test</Socket>);

        expect(component).toMatchSnapshot('socket with rune over');
    });
    test('Renders different splashes', () => {
        const rune: IRune = {
            id: '1',
            image: 'url',
            rarity: Rarity.common,
            type: RuneTypes.attack,
            properties: [
                { name: RunePropertyTypes.Evade, value: 10 },
            ],
        };

        const nodeId = '1';
        let component = shallow(<Socket rune={rune} nodeId={nodeId} isRuneOver>test</Socket>);
        expect(component).toMatchSnapshot('socket with common splash');

        rune.rarity = Rarity.rare;
        component = shallow(<Socket rune={rune} nodeId={nodeId} isRuneOver>test</Socket>);
        expect(component).toMatchSnapshot('socket with rare splash');

        rune.rarity = Rarity.unique;
        component = shallow(<Socket rune={rune} nodeId={nodeId} isRuneOver>test</Socket>);
        expect(component).toMatchSnapshot('socket with unique splash');
    });
    test('Adds animation class', (done) => {
        const rune: IRune = {
            id: '1',
            image: 'url',
            rarity: Rarity.common,
            type: RuneTypes.attack,
            properties: [
                { name: RunePropertyTypes.Evade, value: 10 },
            ],
        };

        const nodeId = '1';
        const component = mount<Socket>(<Socket rune={rune} nodeId={nodeId} isRuneOver>test</Socket>);
        component.instance().animationTimeout = 0;
        const secondRune: IRune = {
            id: '2',
            image: 'url',
            rarity: Rarity.common,
            type: RuneTypes.attack,
            properties: [
                { name: RunePropertyTypes.Evade, value: 10 },
            ],
        };
        component.setProps({
            rune: secondRune,
        });

        expect(component.instance().socketRef.current.classList.contains(utils.constants.splashAnimated)).toEqual(true);

        setTimeout(() => {
            expect(component.instance().socketRef.current.classList.contains(utils.constants.splashAnimated)).toEqual(false);
            done();
        }, 100);
    });
});
