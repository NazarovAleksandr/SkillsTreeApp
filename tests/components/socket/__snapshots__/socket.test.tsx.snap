// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Socket Renders different splashes: socket with common splash 1`] = `
<span
  className="rune-socket drop-allowed common-splash"
>
  test
</span>
`;

exports[`Socket Renders different splashes: socket with rare splash 1`] = `
<span
  className="rune-socket drop-allowed rare-splash"
>
  test
</span>
`;

exports[`Socket Renders different splashes: socket with unique splash 1`] = `
<span
  className="rune-socket drop-allowed unique-splash"
>
  test
</span>
`;

exports[`Socket Renders with child: socket with child 1`] = `
<span
  className="rune-socket  common-splash"
>
  test
</span>
`;

exports[`Socket Renders with rune over: socket with rune over 1`] = `
<span
  className="rune-socket drop-allowed common-splash"
>
  test
</span>
`;
